File: apps/app/src/services/timestamps.ts
import type { UniversalResponse } from '@/types';
import { api } from './api';

export interface Timestamp {
  id: string;
  video_id: string;
  user_id: string;
  timestamp: number;
  title: string;
  note: string;
  tags: string[];
  created_at: string;
  updated_at: string;
}

export interface Tag {
  id: string;
  name: string;
  created_at: string;
  updated_at: string;
}

export interface CreateTimestampRequest {
  video_id: string;
  timestamp: number;
  title?: string;
  note?: string;
  tags?: string[];
}

export interface UpdateTimestampRequest {
  title?: string;
  note?: string;
  tags?: string[];
}

export interface GetTimestampsResponse {
  timestamps: Timestamp[];
  video_id: string;
  count: number;
}

export interface DeleteTimestampResponse {
  message: string;
  timestamp_id: string;
  user_id: string;
}

export interface GetAllTagsResponse {
  tags: Tag[];
  count: number;
}

export interface SearchTagsRequest {
  query: string;
  limit?: number;
}

export interface SearchTagsResponse {
  tags: Tag[];
  count: number;
}

export interface SearchTimestampsRequest {
  query: string;
  video_id?: string;
  limit?: number;
}

export interface GetVideoSummary {
  video_id: string;
  summary: string;
  cached: boolean;
  generated_at: string;
}

export interface SearchTimestampsResponse {
  timestamps: Timestamp[];
  count: number;
}

export interface GenerateFullVideoSummaryRequest {
  video_id: string;
  refresh?: boolean;
}

export interface GenerateFullVideoSummaryResponse {
  summary: string;
  video_id: string;
  video_title: string;
  note_count: number;
  generated_at: string;
  cached: boolean;
}

export interface AnswerQuestionRequest {
  question: string;
  video_id?: string;
  context?: number;
}

export interface AnswerQuestionResponse {
  answer: string;
  question: string;
  video_id?: string;
  relevant_notes: Timestamp[];
  generated_at: string;
}

export const injectedTimestampsApi = api.injectEndpoints({
  endpoints: (builder) => ({
    createTimestamp: builder.mutation<Timestamp, CreateTimestampRequest>({
      query: (data) => ({
        url: '/timestamps',
        method: 'POST',
        body: JSON.stringify(data),
      }),
    }),
    getTimestamps: builder.query<
      UniversalResponse<GetTimestampsResponse>,
      string
    >({
      query: (videoId) => `/timestamps/${videoId}`,
    }),
    getAllTimestamps: builder.query<
      UniversalResponse<{ timestamps: Timestamp[]; count: number }>,
      void
    >({
      query: () => '/timestamps',
    }),
    updateTimestamp: builder.mutation<
      Timestamp,
      { id: string; data: UpdateTimestampRequest }
    >({
      query: ({ id, data }) => ({
        url: `/timestamps/${id}`,
        method: 'PUT',
        body: JSON.stringify(data),
      }),
    }),
    deleteTimestamp: builder.mutation<DeleteTimestampResponse, string>({
      query: (timestampId) => ({
        url: `/timestamps/${timestampId}`,
        method: 'DELETE',
      }),
    }),
    deleteMultipleTimestamps: builder.mutation<{ message: string }, string[]>({
      query: (ids) => ({
        url: '/timestamps',
        method: 'DELETE',
        body: JSON.stringify({ ids }),
      }),
    }),

    // Tag Management
    getAllTags: builder.query<
      UniversalResponse<GetAllTagsResponse>,
      { limit?: number }
    >({
      query: ({ limit = 100 }) => `/timestamps/tags?limit=${limit}`,
    }),
    searchTags: builder.mutation<
      UniversalResponse<SearchTagsResponse>,
      SearchTagsRequest
    >({
      query: (data) => ({
        url: '/timestamps/tags/search',
        method: 'POST',
        body: JSON.stringify(data),
      }),
    }),

    // Search
    searchTimestamps: builder.mutation<
      UniversalResponse<SearchTimestampsResponse>,
      SearchTimestampsRequest
    >({
      query: (data) => ({
        url: '/timestamps/search',
        method: 'POST',
        body: JSON.stringify(data),
      }),
    }),

    // AI Features
    generateFullVideoSummary: builder.mutation<
      UniversalResponse<GenerateFullVideoSummaryResponse>,
      GenerateFullVideoSummaryRequest
    >({
      query: (data) => ({
        url: '/timestamps/full-summary',
        method: 'POST',
        body: JSON.stringify(data),
      }),
    }),
    getVideoSummary: builder.query<UniversalResponse<GetVideoSummary>, string>({
      query: (videoId) => `/timestamps/summary/${videoId}`,
    }),

    answerQuestion: builder.mutation<
      UniversalResponse<AnswerQuestionResponse>,
      AnswerQuestionRequest
    >({
      query: (data) => ({
        url: '/timestamps/question',
        method: 'POST',
        body: JSON.stringify(data),
      }),
    }),
  }),
});

export const {
  useCreateTimestampMutation,
  useGetTimestampsQuery,
  useGetAllTimestampsQuery,
  useUpdateTimestampMutation,
  useDeleteTimestampMutation,
  useDeleteMultipleTimestampsMutation,
  useGetAllTagsQuery,
  useSearchTagsMutation,
  useSearchTimestampsMutation,
  useGenerateFullVideoSummaryMutation,
  useAnswerQuestionMutation,
  useGetVideoSummaryQuery,
} = injectedTimestampsApi;

File: backend/internal/models/timestamps.go
package models

import "time"

type Tag struct {
	ID        string    `json:"id" bun:"id,pk"`
	Name      string    `json:"name" bun:"name,notnull,unique"`
	CreatedAt time.Time `json:"created_at" bun:"created_at,notnull"`
	UpdatedAt time.Time `json:"updated_at" bun:"updated_at,notnull"`
}

func (Tag) TableName() string {
	return "tags"
}

type Timestamp struct {
	ID        string    `json:"id" bun:"id,pk"`
	VideoID   string    `json:"video_id" bun:"video_id,notnull"`
	UserID    string    `json:"user_id" bun:"user_id,notnull"`
	Timestamp float64   `json:"timestamp" bun:"timestamp,notnull"`
	Title     string    `json:"title"`
	Note      string    `json:"note"`
	Tags      []Tag     `json:"tags" bun:"m2m:timestamp_tags,join:Timestamp=Tag"`
	Embedding []float32 `json:"-" bun:"embedding,notnull"`
	CreatedAt time.Time `json:"created_at" bun:"created_at,notnull"`
	UpdatedAt time.Time `json:"updated_at" bun:"updated_at,notnull"`
	DeletedAt time.Time `json:"-" bun:"deleted_at,soft_delete,nullzero"`
}

func (Timestamp) TableName() string {
	return "timestamps"
}

type TimestampTag struct {
	ID          string    `json:"id" bun:"id,pk"`
	TimestampID string    `json:"timestamp_id" bun:"timestamp_id,notnull"`
	TagID       string    `json:"tag_id" bun:"tag_id,notnull"`
	CreatedAt   time.Time `json:"created_at" bun:"created_at,notnull"`
}

func (TimestampTag) TableName() string {
	return "timestamp_tags"
}
